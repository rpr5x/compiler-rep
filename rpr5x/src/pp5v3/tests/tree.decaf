

class TreeItem {
  int data;
  TreeItem left;
  TreeItem right;
  
  TreeItem Init(int data) {
    this.data = data;
    left = null;
    right = null;
    return this;
  }
  int GetData() { return data;  }
  TreeItem GetLeft() { return left;}
  TreeItem GetRight() { return right;}
  void SetLeft(TreeItem left) { this.left = left; }
  void SetRight(TreeItem right) { this.right = right; }
   void SetData(int data) { this.data = data;}
  void PrintNode() {
      if (GetLeft() != null) GetLeft().PrintNode();
       Print(" ", data, " ");
       if (GetRight() != null) GetRight().PrintNode();
  }  
}  


class Tree {
  TreeItem root;
  
  void Init() {
    root = null;
  }

  void Helper(int i, TreeItem node) {
    if (i <= node.GetData()) {
      if (node.GetLeft() == null) {
        TreeItem l;
        l = New(TreeItem);
	  node.SetLeft(l.Init(i));
      } else {
	Helper(i, node.GetLeft());	
      }
    } else {
      if (node.GetRight() == null) {
         TreeItem r;
         r = New(TreeItem);
         node.SetRight(r.Init(i));
      } else {
	Helper(i, node.GetRight());
      }
    }
  }

  void Insert(int i) {
    if (root == null) {
      root = New(TreeItem);
      root.Init(i);
    } else {
      Helper(i, root);
    }
  }

  void PrintTree() {
	if (root != null) root.PrintNode();
      else Print("<empty>");
     Print("\n");
 }
}

void main() {
  Tree t;

  t = New(Tree);
  t.PrintTree();
  t.Insert(5);
  t.Insert(10);
   t.Insert(0);
   t.PrintTree();
   t.Insert(10);
   t.Insert(4);
   t.Insert(17);
  t.Insert(8);
   t.Insert(6);
  t.Insert(3);
  t.PrintTree();
}
