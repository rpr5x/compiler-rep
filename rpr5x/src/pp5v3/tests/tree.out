Loaded: /if14/ww6r/WorkingDisk/TA/compiler/exceptions/exceptions.s
<empty>
 0  5  10 
 0  3  4  5  6  8  10  10  17 
